File: Backend/api/schemas/budget.py
from __future__ import annotations

import datetime as dt
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field


class BudgetCreate(BaseModel):
    amount: Decimal = Field(gt=0)
    month: Optional[int] = Field(default=None, ge=1, le=12)
    year: Optional[int] = Field(default=None, ge=2000, le=2100)
    name: Optional[str] = Field(default=None, max_length=100)
    alert_threshold: Optional[float] = Field(default=None, ge=0, le=1)


class BudgetRead(BaseModel):
    id: str
    name: str
    month: int
    year: int
    amount: float
    currency: str
    spent: float
    remaining: float
    alert_threshold: float
    created_at: dt.datetime
    updated_at: dt.datetime

    class Config:
        from_attributes = True

File: mobile/lib/data/models/budget.dart
class Budget {
  Budget({
    required this.id,
    required this.name,
    required this.month,
    required this.year,
    required this.amount,
    required this.currency,
    required this.spent,
    required this.remaining,
    required this.alertThreshold,
    required this.updatedAt,
  });

  final String id;
  final String name;
  final int month;
  final int year;
  final double amount;
  final String currency;
  final double spent;
  final double remaining;
  final double alertThreshold;
  final DateTime updatedAt;

  factory Budget.fromJson(Map<String, dynamic> json) {
    return Budget(
      id: json['id'] as String,
      name: json['name'] as String? ?? 'Presupuesto',
      month: json['month'] as int,
      year: json['year'] as int,
      amount: (json['amount'] as num).toDouble(),
      currency: (json['currency'] as String?) ?? 'PEN',
      spent: (json['spent'] as num).toDouble(),
      remaining: (json['remaining'] as num).toDouble(),
      alertThreshold: (json['alert_threshold'] as num).toDouble(),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }
}
